usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -0,0 +1,291 @@
+    direct_payment_type_mapping,
+    grant_subaward_mapping,
+    grant_type_mapping,
+    idv_type_mapping,
+    loan_type_mapping,
+    non_loan_assistance_type_mapping,
+    other_type_mapping,
+    procurement_type_mapping,
+)
+from usaspending_api.awards.v2.lookups.matview_lookups import (
+    award_contracts_mapping,
+    award_idv_mapping,
+    loan_award_mapping,
+    non_loan_assistance_award_mapping,
+)
+from usaspending_api.common.api_versioning import api_transformations, API_TRANSFORM_FUNCTIONS
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.exceptions import InvalidParameterException, UnprocessableEntityException
+from usaspending_api.common.helpers.orm_helpers import (
+    obtain_view_from_award_group,
+    award_types_are_valid_groups,
+    subaward_types_are_valid_groups,
+)
+from usaspending_api.common.validator.award_filter import AWARD_FILTER
+from usaspending_api.common.validator.pagination import PAGINATION
+from usaspending_api.common.validator.tinyshield import TinyShield
+
+
+@api_transformations(api_version=settings.API_VERSION, function_list=API_TRANSFORM_FUNCTIONS)
+class SpendingByAwardVisualizationViewSet(APIView):
+    """
+    This route takes award filters and fields, and returns the fields of the filtered awards.
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/search/AdvancedSearch.md"
+
+    @cache_response()
+    def post(self, request):
+        """Return all awards matching the provided filters and limits"""
+
+        json_request = self.validate_request_data(request.data)
+        filters = json_request["filters"]
+        subawards = json_request["subawards"]
+        self.fields = json_request["fields"]
+        self.pagination = {
+            "limit": json_request["limit"],
+            "lower_bound": (json_request["page"] - 1) * json_request["limit"],
+            "page": json_request["page"],
+            "sort_key": json_request.get("sort") or self.fields[0],
+            "sort_order": json_request["order"],
+            "upper_bound": json_request["page"] * json_request["limit"] + 1,
+        }
+
+        self._response_dict = {
+            "limit": self.pagination["limit"],
+            "results": None,
+            "page_metadata": {"page": self.pagination["page"], "hasNext": None},
+        }
+
+        if "no intersection" in filters["award_type_codes"]:
+            # "Special case": there will never be results when the website provides this value
+            self.populate_response(results=[], has_next=False)
+            return Response(self._response_dict)
+
+        self.raise_if_award_types_not_valid_subset(filters["award_type_codes"], subawards)
+        self.raise_if_sort_key_not_valid(self.pagination["sort_key"], self.fields, subawards)
+
+        if subawards:
+            self._construct_subaward_variables(filters)
+        else:
+            self._construct_award_variables(filters)
```

Comment:
Is it unpythonic to have private functions return a value that can then be set as a variable, rather than change around class variables behind the scenes? Referential Transparency isn't a principle here, but as a future dev I wouldn't know when I would and wouldn't want to dive into this function.

Reply:
Did you see how many class attributes are being set in those methods?

Reply:
the section between 118 and 130 is an example of a bunch of things changing at once, some in scope of the function, some in scope of the class.